### PR TITLE
find-runtime-roots: fold into gc.cc

### PR DESCRIPTION
**Pull request by shlevy**

### CODE REVIEW

File: src/libstore/gc.cc [hunk 1]
```diff
@@ -330,18 +335,184 @@ Roots LocalStore::findRoots()
...
+    std::remove_reference<decltype(errno)>::type child_errno;
+    auto report_read_count = size_t{0};
+    while (report_read_count < sizeof child_errno) {
+        auto res = read(lsof_report.readSide.get(), reinterpret_cast<unsigned char *>(&child_errno) + report_read_count, sizeof child_errno - report_read_count);
+        if (res == -1)
+            throw SysError("reading from lsof report pipe");
+        if (res == 0)
+            break;
+        report_read_count += res;
+    }
```
Comment (edolstra) [inline src/libstore/gc.cc]: Why not use `runProgram()`? That should cut a few dozen lines of code.

File: src/libstore/gc.cc [hunk 2]
```diff
@@ -330,18 +335,184 @@ Roots LocalStore::findRoots()
...
+                if (errno)
+                    throw SysError(format("iterating /proc/%1%/fd") % ent->d_name);
+                fdDir.close();
+
+                auto mapFd = AutoCloseFD{openat(procFd.get(), "maps", O_RDONLY | O_CLOEXEC)};
+                if (!mapFd) {
+                    if (errno == ENOENT || errno == EACCES)
+                        continue;
+                    throw SysError(format("opening /proc/%1%/maps") % ent->d_name);
+                }
```
Comment (edolstra) [inline src/libstore/gc.cc]: => `readFile()` (with `drain=true`).

File: src/libstore/gc.cc [hunk 3]
```diff
@@ -330,18 +335,184 @@ Roots LocalStore::findRoots()
 }
 
 
+static void readProcLink(int procFd, const char * file, StringSet & paths)
```
Comment (edolstra) [inline src/libstore/gc.cc]: Better to just pass a full path here rather than a file descriptor (e.g. don't use `*at()` system calls).

Comment (edolstra) [inline src/libstore/gc.cc]: And then `readProcLink()` would become `readLink()`.

Comment (shlevy) [inline src/libstore/gc.cc]: Why is it better?

File: src/libstore/gc.cc (hunk 2 above)
Comment (shlevy) [inline src/libstore/gc.cc]: Doesn't that read the whole file into memory at once?

File: src/libstore/gc.cc (hunk 1 above)
Comment (shlevy) [inline src/libstore/gc.cc]: Can't put stderr to dev null, can't differentiate error in lsof (which is a problem) from lsof not being installed.

Comment (shlevy) [inline src/libstore/gc.cc]: Oh, and have to read the whole output into memory at once

Comment (edolstra) [inline src/libstore/gc.cc]: Well, we can add some more options to `runProgram()`. Reading the entire output into memory doesn't seem problematic.

File: src/libstore/gc.cc (hunk 2 above)
Comment (edolstra) [inline src/libstore/gc.cc]: Yes, but that shouldn't be a problem.

File: src/libstore/gc.cc (hunk 3 above)
Comment (edolstra) [inline src/libstore/gc.cc]: Because it's a lot simpler than all this passing directory file descriptors around.

Comment (shlevy) [inline src/libstore/gc.cc]: @edolstra I can't use `readLink` here because `lstat` on files in proc always returns 0 for size.

**Comment by shlevy** [comment]
@edolstra updated

**Comment by domenkozar** [comment]
@shlevy PR has conflicts.

**Comment by shlevy** [comment]
Will update when I have a go-ahead for merging.

**Comment by shlevy** [comment]
(it's a trivial conflict in local.mk)

**Pull request merged**